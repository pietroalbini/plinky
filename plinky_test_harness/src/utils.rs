use anyhow::{bail, Error};
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn run(command: &mut Command) -> Result<(), Error> {
    let cmd_repr = format!("{command:?}");
    let output = command.output()?;
    if !output.status.success() {
        eprintln!("Failed to execute {cmd_repr}");
        eprintln!();
        eprintln!("=== stdout ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stdout));
        eprintln!();
        eprintln!("=== stderr ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        eprintln!();
        bail!("command failed with exit {}", output.status);
    }
    Ok(())
}

pub struct RunAndSnapshot {
    name: String,
    path: PathBuf,
    output: String,
}

impl RunAndSnapshot {
    pub fn new(name: &str, path: &Path) -> Self {
        Self { name: name.into(), path: path.into(), output: String::new() }
    }

    pub fn run(&mut self, action: &str, command: &mut Command) -> Result<bool, Error> {
        self.separator();
        match command.output() {
            Ok(output) => {
                self.output.push_str(&format!("{action} exited with {}\n", output.status));
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.trim_ascii().is_empty() {
                        self.output.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let mut content = String::from_utf8_lossy(content).to_string();
                        content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        self.output
                            .push_str(&format!("\n=== {name} ===\n{}\n", content.trim_ascii_end()));
                    }
                }
                Ok(output.status.success())
            }
            Err(err) => {
                self.output.push_str(&format!("{action} failed to execute with error: {err}"));
                Ok(false)
            }
        }
    }

    pub fn note(&mut self, note: &str) {
        self.separator();
        self.output.push_str(note);
        if !note.ends_with('\n') {
            self.output.push('\n');
        }
    }

    fn separator(&mut self) {
        if !self.output.is_empty() {
            self.output.push_str("\n==============\n\n");
        }
    }

    pub fn persist(self) {
        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(self.path.canonicalize().unwrap());

        insta_settings.bind(|| {
            insta::assert_snapshot!(self.name, self.output.trim_ascii_end());
        });
    }
}

pub(crate) fn file_name(path: impl AsRef<Path>) -> String {
    path.as_ref().file_name().unwrap().to_str().unwrap().to_string()
}

#[track_caller]
pub(crate) fn err_str<T>(result: Result<T, Error>) -> Result<T, String> {
    match result {
        Ok(ok) => Ok(ok),
        Err(err) => {
            let mut repr = format!("error: {err}\n");
            let mut source = err.source();
            while let Some(err) = source {
                repr.push_str(&format!("  cause: {err}\n"));
                source = err.source();
            }
            panic!("{repr}")
        }
    }
}
