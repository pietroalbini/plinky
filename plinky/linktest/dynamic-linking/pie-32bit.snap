---
source: plinky_test_harness/src/utils.rs
---
linking exited with exit status: 0

no stdout present

=== stderr ===
debug print: loaded object
 │
 │  section .text (perms: rx) in entry.o
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 55 89 e5 53 83 ec 04 e8 fc ff ff ff 05 01 00 00 │ U..S............ │
 │   │  │ 00 89 c3 e8 fc ff ff ff 90 8b 5d fc c9 c3       │ ..........]...   │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   │
 │   │    Relocations:
 │   │  ╭───────────────────────┬───────────────────────┬────────┬──────────╮
 │   │  │ Type                  │ Symbol                │ Offset │ Addend   │
 │   │  ├───────────────────────┼───────────────────────┼────────┼──────────┤
 │   │  │ Relative32            │ __x86.get_pc_thunk.ax │ 0x8    │ <inline> │
 │   │  ├───────────────────────┼───────────────────────┼────────┼──────────┤
 │   │  │ GOTLocationRelative32 │ _GLOBAL_OFFSET_TABLE_ │ 0xd    │ <inline> │
 │   │  ├───────────────────────┼───────────────────────┼────────┼──────────┤
 │   │  │ PLT32                 │ impl                  │ 0x14   │ <inline> │
 │   │  ╰───────────────────────┴───────────────────────┴────────┴──────────╯
 │   ┴
 │
 │    Symbols:
 │  ╭───────────────────────────────────────┬──────────┬────────────┬─────────────────┬───────────────────────────────────╮
 │  │ Name                                  │ Type     │ Source     │ Visibility      │ Value                             │
 │  ├───────────────────────────────────────┼──────────┼────────────┼─────────────────┼───────────────────────────────────┤
 │  │ <empty>                               │ none     │ <plinky>   │ local           │ <null>                            │
 │  ├───────────────────────────────────────┼──────────┼────────────┼─────────────────┼───────────────────────────────────┤
 │  │ <section .text>                       │ section  │ entry.o    │ local           │ .text + 0x0                       │
 │  ├───────────────────────────────────────┼──────────┼────────────┼─────────────────┼───────────────────────────────────┤
 │  │ <section .text.__x86.get_pc_thunk.ax> │ section  │ entry.o    │ local           │ .text.__x86.get_pc_thunk.ax + 0x0 │
 │  ├───────────────────────────────────────┼──────────┼────────────┼─────────────────┼───────────────────────────────────┤
 │  │ _GLOBAL_OFFSET_TABLE_                 │ none     │ entry.o    │ global          │ <undefined>                       │
 │  ├───────────────────────────────────────┼──────────┼────────────┼─────────────────┼───────────────────────────────────┤
 │  │ __x86.get_pc_thunk.ax                 │ function │ entry.o    │ global (hidden) │ .text.__x86.get_pc_thunk.ax + 0x0 │
 │  ├───────────────────────────────────────┼──────────┼────────────┼─────────────────┼───────────────────────────────────┤
 │  │ _start                                │ function │ entry.o    │ global          │ .text + 0x0                       │
 │  ├───────────────────────────────────────┼──────────┼────────────┼─────────────────┼───────────────────────────────────┤
 │  │ impl                                  │ none     │ libimpl.so │ global          │ <externally defined>              │
 │  ╰───────────────────────────────────────┴──────────┴────────────┴─────────────────┴───────────────────────────────────╯
 │
 │  inputs
 │   │
 │   │    entry.o
 │   │  ╭─────────────────────┬─────╮
 │   │  │ X86 ISA used        │     │
 │   │  ├─────────────────────┼─────┤
 │   │  │ x86 features 2 used │ x86 │
 │   │  ╰─────────────────────┴─────╯
 │   │
 │   │  libimpl.so (shared object)
 │   ┴
 ┴

debug print: built elf
 │
 │  section .dynstr (address: 0x174)
 │   │
 │   │    Strings table:
 │   │  ╭─────┬────────────╮
 │   │  │ 0x0 │            │
 │   │  ├─────┼────────────┤
 │   │  │ 0x1 │ libimpl.so │
 │   │  ├─────┼────────────┤
 │   │  │ 0xc │            │
 │   │  ├─────┼────────────┤
 │   │  │ 0xd │ impl       │
 │   │  ╰─────┴────────────╯
 │   ┴
 │
 │  section .dynsym (address: 0x186)
 │   │
 │   │    Dynamic symbol table:
 │   │  ╭───────────┬─────────┬──────┬────────────┬────────────┬───────┬──────╮
 │   │  │ Name      │ Binding │ Type │ Visibility │ Definition │ Value │ Size │
 │   │  ├───────────┼─────────┼──────┼────────────┼────────────┼───────┼──────┤
 │   │  │ <empty>#1 │ Local   │ -    │ Default    │ Undefined  │ 0x0   │ 0x0  │
 │   │  ├───────────┼─────────┼──────┼────────────┼────────────┼───────┼──────┤
 │   │  │ impl#1    │ Global  │ -    │ Default    │ Undefined  │ 0x0   │ 0x0  │
 │   │  ╰───────────┴─────────┴──────┴────────────┴────────────┴───────┴──────╯
 │   ┴
 │
 │  section .dynamic (address: 0x1001c2)
 │   │
 │   │  dynamic information | string table: .dynstr
 │   │
 │   │  ╭─────────────────────────┬──────────────────────────────╮
 │   │  │ Kind                    │ Value                        │
 │   │  ├─────────────────────────┼──────────────────────────────┤
 │   │  │ Needed libraries        │ offset 1 in the string table │
 │   │  ├─────────────────────────┼──────────────────────────────┤
 │   │  │ String table            │ address 0x174                │
 │   │  ├─────────────────────────┼──────────────────────────────┤
 │   │  │ String table size       │ 18 bytes                     │
 │   │  ├─────────────────────────┼──────────────────────────────┤
 │   │  │ Symbol table            │ address 0x186                │
 │   │  ├─────────────────────────┼──────────────────────────────┤
 │   │  │ Symbol table entry size │ 16 bytes                     │
 │   │  ├─────────────────────────┼──────────────────────────────┤
 │   │  │ Hash table              │ address 0x1a6                │
 │   │  ├─────────────────────────┼──────────────────────────────┤
 │   │  │ Jump PLT relocations    │ address 0x100222             │
 │   │  ├─────────────────────────┼──────────────────────────────┤
 │   │  │ PLT relocations size    │ 8 bytes                      │
 │   │  ├─────────────────────────┼──────────────────────────────┤
 │   │  │ PLT relocations type    │ Relocations                  │
 │   │  ├─────────────────────────┼──────────────────────────────┤
 │   │  │ PLT GOT                 │ address 0x103000             │
 │   │  ├─────────────────────────┼──────────────────────────────┤
 │   │  │ Flags1                  │ pie                          │
 │   │  ├─────────────────────────┼──────────────────────────────┤
 │   │  │ Null                    │ -                            │
 │   │  ╰─────────────────────────┴──────────────────────────────╯
 │   ┴
 ┴

==============

running exited with exit status: 0

no stdout present

no stderr present
